hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java

```diff
@@ -400,39 +391,34 @@ public void reset() {
     public CompletableFuture<JobMetrics> getJobMetrics(long jobId) {
         CompletableFuture<JobMetrics> cf = new CompletableFuture<>();
         submitToCoordinatorThread(
-                () -> {
-                    // first check if there is a job result present.
-                    // this map is updated first during completion.
-                    JobMetrics metrics = jobRepository.getJobMetrics(jobId);
-                    if (metrics != null) {
-                        cf.complete(metrics);
-                        return;
-                    }
-
-                    // check if there is a master context for the running job
-                    MasterContext mc = masterContexts.get(jobId);
-                    if (mc != null) {
-                        mc.jobContext().collectMetrics(cf);
-                        return;
-                    }
+            () -> {
+                // check if there is a running job
+                MasterContext mc = masterContexts.get(jobId);
+                if (mc != null) {
+                    mc.jobContext().collectMetrics(cf);
+                    return;
+                }
 
-                    // no master context found, job might be just submitted
-                    JobExecutionRecord jobExecutionRecord = jobRepository.getJobExecutionRecord(jobId);
-                    if (jobExecutionRecord != null) {
-                        cf.complete(JobMetrics.empty());
-                        return;
-                    } else {
-                        // no job record found, but check job results again
-                        // since job might have been completed meanwhile.
-                        metrics = jobRepository.getJobMetrics(jobId);
-                        if (metrics != null) {
-                            cf.complete(metrics);
-                            return;
-                        }
-                    }
+                JobExecutionRecord record = jobRepository.getJobExecutionRecord(jobId);
+                if (record != null) {
+                    // job is suspended, or just submitted now
+                    cf.complete(JobMetrics.empty());
+                    return;
+                }
 
-                    cf.completeExceptionally(new JobNotFoundException(jobId));
+                JobResult jobResult = jobRepository.getJobResult(jobId);
```

Comment:
Also we can do this the other way: first check JobMetrics, if not found, then check JobResult before throwing. It's unlikely the user will check for jobMetrics if he disabled it.